src/network/wait-online/manager.c

```diff
@@ -381,13 +406,122 @@ static int manager_network_monitor_listen(Manager *m) {
         return 0;
 }
 
+static int on_dns_configuration_event(
+                sd_varlink *link,
+                sd_json_variant *parameters,
+                const char *error_id,
+                sd_varlink_reply_flags_t flags,
+                void *userdata) {
+
+        Manager *m = ASSERT_PTR(userdata);
+        sd_json_variant *configuration = NULL, *v = NULL;
+        int r;
+
+        assert(link);
+
+        if (error_id) {
+                log_warning("DNS configuration event error, ignoring: %s", error_id);
+                return 0;
+        }
+
+        configuration = sd_json_variant_by_key(parameters, "configuration");
+        if (!configuration || !sd_json_variant_is_array(configuration)) {
+                log_warning("DNS configuration JSON data does not have configuration key, ignoring.");
+                return 0;
+        }
+
+        JSON_VARIANT_ARRAY_FOREACH(v, configuration) {
+                _cleanup_strv_free_ char **dns = NULL;
+                sd_json_variant *w = NULL;
+                uint64_t ifindex;
+
+                if (!sd_json_variant_is_object(v)) {
+                        log_warning("DNS event JSON is not an array of objects, ignoring.");
+                        continue;
+                }
+
+                w = sd_json_variant_by_key(v, "interfaceIndex");
```

Comment:
these days we prefer if sd_json_dispatch() is used to map json fields to a C struct

Reply:
Thanks, I have revised this.